File: useAlgorithm.py
import random
import numpy as np

class Student:
    def __init__(self, name, preferences, avoids):
        self.name = name
        self.preferences = preferences
        self.avoids = avoids

class Table:
    def __init__(self, size):
        self.size = size
        self.students = []

    def add_student(self, student):
        if len(self.students) < self.size and student not in self.students:
            self.students.append(student)
            return True
        return False

    def remove_student(self, student):
        if student in self.students:
            self.students.remove(student)
            return True
        return False

    def __repr__(self):
        return f"Table({[student.name for student in self.students]})"

def calculate_gini_coefficient(data):
    sorted_data = sorted(data)
    n = len(data)
    cumulative_data = np.cumsum(sorted_data)
    return (2 * np.sum((i + 1) * sorted_data[i] for i in range(n)) - (n + 1) * cumulative_data[-1]) / (n * cumulative_data[-1])

def count_satisfied_preferences(tables):
    score = 0
    preference_counts = []
    
    for table in tables:
        table_preference_count = 0
        for student in table.students:
            for preference in student.preferences:
                if preference in [s.name for s in table.students]:
                    score += 1
                    table_preference_count += 1
        preference_counts.append(table_preference_count)
    
    for table in tables:
        for student in table.students:
            for avoid in student.avoids:
                if avoid in [s.name for s in table.students]:
                    score -= 1
    
    gini_coefficient = calculate_gini_coefficient(preference_counts)
    
    if gini_coefficient >= 0 and gini_coefficient < 0.5:
        reward = 10 - round(gini_coefficient, 2) * 10
        score += reward
    else:
        penalty = round(gini_coefficient, 2) * 10
        score -= penalty
    
    return score, gini_coefficient

def assign_students_to_tables(students, num_tables, table_size):
    tables = [Table(table_size) for _ in range(num_tables)]
    random.shuffle(students)
    
    for student in students:
        best_table = None
        max_preferences = -1
        
        for table in tables:
            current_preferences = sum(1 for s in table.students if s.name in student.preferences)
            
            if len(table.students) < table.size and current_preferences > max_preferences:
                best_table = table
                max_preferences = current_preferences
        
        if best_table:
            best_table.add_student(student)
    
    return tables

def optimize_seating(tables, iterations=5000):
    best_tables = tables
    best_score = count_satisfied_preferences(tables)[0]
    
    for _ in range(iterations):
        table1, table2 = random.sample(tables, 2)
        if not table1.students or not table2.students:
            continue

        student1 = random.choice(table1.students)
        student2 = random.choice(table2.students)

        table1.remove_student(student1)
        table2.remove_student(student2)
        
        table1.add_student(student2)
        table2.add_student(student1)
        
        new_score = count_satisfied_preferences(tables)[0]
        if new_score > best_score:
            best_score = new_score
            best_tables = [Table(t.size) for t in tables]
            for t, bt in zip(tables, best_tables):
                bt.students = t.students.copy()
        else:
            table1.remove_student(student2)
            table2.remove_student(student1)
            table1.add_student(student1)
            table2.add_student(student2)
    
    return best_tables

def find_best_seating(students, num_tables, table_size, trials=10):
    best_seating = None
    best_score = float('-inf')
    
    for _ in range(trials):
        tables = assign_students_to_tables(students, num_tables, table_size)
        tables = optimize_seating(tables)
        score = count_satisfied_preferences(tables)[0]
        if score > best_score:
            best_score = score
            best_seating = tables
    
    return best_seating

students = [
    Student('Alice', ['Bob', 'Charlie', 'David'], []),
    Student('Bob', ['Alice'], ['Yara', 'Tina']),
    Student('Charlie', ['Alice', 'Uma', 'Chloe'], []),
    Student('David', ['Alice', 'Bob', 'Charlie'], []),
    Student('Eve', ['Frank', 'Grace', 'Hannah', 'Ivy'], ['Paul', 'Charlie']),
    Student('Frank', ['Alice', 'Bob', 'Tina', 'Jack'], ['Kate']),
    Student('Grace', ['Charlie'], ['Violet']),
    Student('Hannah', ['Jack', 'Uma', 'Chloe', 'Harry'], ['Erick']),
    Student('Ivy', ['Charlie'], []),
    Student('Jack', ['Bob', 'Alice', 'David'], ['Tina', 'Ruby']),
    Student('Kate', ['Frank', 'Leo', 'Gina'], []),
    Student('Leo', ['Isla', 'Jack', 'Frank', 'Ivy'], []),
    Student('Mia', ['Ruby', 'Uma'], []),
    Student('Nina', ['Bob'], ['Ava']),
    Student('Oscar', ['Violet'], []),
    Student('Paul', ['Grace'], ['Charlie']),
    Student('Quinn', ['Tina'], []),
    Student('Ruby', ['Hannah', 'Grace', 'Kate'], []),
    Student('Sam', ['Tina', 'Gina', 'Paul', 'Kate'], []),
    Student('Tina', ['Yara', 'Harry'], []),
    Student('Uma', ['Alice'], []),
    Student('Violet', ['Uma', 'Oscar'], ['Frank', 'Ivy']),
    Student('Will', ['Gina', 'Eve', 'Quinn', 'Alice'], []),
    Student('Xander', ['Gina', 'Ruby'], []),
    Student('Yara', ['Ben', 'Hannah'], []),
    Student('Zack', ['Nina', 'Paul', 'Frank'], []),
    Student('Ava', ['Paul', 'Sam', 'Oscar', 'Xander'], []),
    Student('Ben', ['Ava', 'Harry', 'Nina', 'Xander'], ['Ava']),
    Student('Chloe', ['Sam', 'Ben'], ['Gina', 'Quinn']),
    Student('Dylan', ['Hannah', 'Bob'], []),
    Student('Ellie', ['Quinn', 'Alice', 'Tina'], []),
    Student('Finn', ['Mia', 'Jack', 'Ellie', 'Chloe'], []),
    Student('Gina', ['Nina', 'Tina', 'Jack'], ['Ruby']),
    Student('Harry', ['Sam', 'Nina'], []),
    Student('Isla', ['Ava', 'David', 'Violet'], []),
    Student('Erick', ['Ben', 'Xander'], [])
]

num_tables = 9
table_size = 4

best_seating = find_best_seating(students, num_tables, table_size, trials=10)
for table in best_seating:
    print(table)
print(f"Best Score: {count_satisfied_preferences(best_seating)[0]}")
print(f"Gini Coefficient: {count_satisfied_preferences(best_seating)[1]}")
      

File: useAlgorithm.js
class Student {
    constructor(name, preferences, avoids) {
        this.name = name;
        this.preferences = preferences;
        this.avoids = avoids;
    }
}

class Table {
    constructor(size) {
        this.size = size;
        this.students = [];
    }

    addStudent(student) {
        if (this.students.length < this.size && !this.students.includes(student)) {
            this.students.push(student);
            return true;
        }
        return false;
    }

    removeStudent(student) {
        const index = this.students.indexOf(student);
        if (index !== -1) {
            this.students.splice(index, 1);
            return true;
        }
        return false;
    }

    toString() {
        return `Table(${this.students.map(student => student.name)})`;
    }
}

function calculateGiniCoefficient(data) {
    const sortedData = data.slice().sort((a, b) => a - b);
    const n = data.length;
    const cumulativeData = sortedData.reduce((acc, val) => {
        acc.push((acc.length > 0 ? acc[acc.length - 1] : 0) + val);
        return acc;
    }, []);
    const sumOfProducts = sortedData.reduce((acc, val, i) => acc + (i + 1) * val, 0);
    const cumulativeSum = cumulativeData[cumulativeData.length - 1];
    return (2 * sumOfProducts - (n + 1) * cumulativeSum) / (n * cumulativeSum);
}

function countSatisfiedPreferences(tables) {
    let score = 0;
    const preferenceCounts = [];
    
    for (const table of tables) {
        let tablePreferenceCount = 0;
        for (const student of table.students) {
            for (const preference of student.preferences) {
                if (table.students.some(s => s.name === preference)) {
                    score++;
                    tablePreferenceCount++;
                }
            }
        }
        preferenceCounts.push(tablePreferenceCount);
    }
    
    for (const table of tables) {
        for (const student of table.students) {
            for (const avoid of student.avoids) {
                if (table.students.some(s => s.name === avoid)) {
                    score--;
                }
            }
        }
    }

    const giniCoefficient = calculateGiniCoefficient(preferenceCounts);

    if (giniCoefficient >= 0 && giniCoefficient < 0.5) {
        const reward = 10 - Math.round(giniCoefficient * 10 * 100) / 100;
        score += reward;
    } else {
        const penalty = Math.round(giniCoefficient * 10 * 100) / 100;
        score -= penalty;
    }

    return { score, giniCoefficient };
}

function assignStudentsToTables(students, numTables, tableSize) {
    const tables = Array.from({ length: numTables }, () => new Table(tableSize));
    students = students.slice();
    students.sort(() => Math.random() - 0.5);
    
    for (const student of students) {
        let bestTable = null;
        let maxPreferences = -1;
        
        for (const table of tables) {
            const currentPreferences = table.students.filter(s => student.preferences.includes(s.name)).length;
            if (table.students.length < table.size && currentPreferences > maxPreferences) {
                bestTable = table;
                maxPreferences = currentPreferences;
            }
        }

        if (bestTable) {
            bestTable.addStudent(student);
        }
    }

    return tables;
}

function optimizeSeating(tables, iterations = 5000) {
    let bestTables = tables.map(t => new Table(t.size));
    tables.forEach((table, i) => bestTables[i].students = table.students.slice());
    let bestScore = countSatisfiedPreferences(tables).score;

    for (let i = 0; i < iterations; i++) {
        const [table1, table2] = tables.slice().sort(() => Math.random() - 0.5).slice(0, 2);
        if (table1.students.length === 0 || table2.students.length === 0) continue;

        const student1 = table1.students[Math.floor(Math.random() * table1.students.length)];
        const student2 = table2.students[Math.floor(Math.random() * table2.students.length)];

        table1.removeStudent(student1);
        table2.removeStudent(student2);
        
        table1.addStudent(student2);
        table2.addStudent(student1);

        const newScore = countSatisfiedPreferences(tables).score;
        if (newScore > bestScore) {
            bestScore = newScore;
            bestTables = tables.map(t => new Table(t.size));
            tables.forEach((table, i) => bestTables[i].students = table.students.slice());
        } else {
            table1.removeStudent(student2);
            table2.removeStudent(student1);
            table1.addStudent(student1);
            table2.addStudent(student2);
        }
    }

    return bestTables;
}

function findBestSeating(students, numTables, tableSize, trials = 10) {
    let bestSeating = null;
    let bestScore = -Infinity;

    for (let i = 0; i < trials; i++) {
        let tables = assignStudentsToTables(students, numTables, tableSize);
        tables = optimizeSeating(tables);
        const { score } = countSatisfiedPreferences(tables);
        if (score > bestScore) {
            bestScore = score;
            bestSeating = tables;
        }
    }

    return bestSeating;
}

const students = [
    new Student('Alice', ['Bob', 'Charlie', 'David'], []),
    new Student('Bob', ['Alice'], ['Yara', 'Tina']),
    new Student('Charlie', ['Alice', 'Uma', 'Chloe'], []),
    new Student('David', ['Alice', 'Bob', 'Charlie'], []),
    new Student('Eve', ['Frank', 'Grace', 'Hannah', 'Ivy'], ['Paul', 'Charlie']),
    new Student('Frank', ['Alice', 'Bob', 'Tina', 'Jack'], ['Kate']),
    new Student('Grace', ['Charlie'], ['Violet']),
    new Student('Hannah', ['Jack', 'Uma', 'Chloe', 'Harry'], ['Erick']),
    new Student('Ivy', ['Charlie'], []),
    new Student('Jack', ['Bob', 'Alice', 'David'], ['Tina', 'Ruby']),
    new Student('Kate', ['Frank', 'Leo', 'Gina'], []),
    new Student('Leo', ['Isla', 'Jack', 'Frank', 'Ivy'], []),
    new Student('Mia', ['Ruby', 'Uma'], []),
    new Student('Nina', ['Bob'], ['Ava']),
    new Student('Oscar', ['Violet'], []),
    new Student('Paul', ['Grace'], ['Charlie']),
    new Student('Quinn', ['Tina'], []),
    new Student('Ruby', ['Hannah', 'Grace', 'Kate'], []),
    new Student('Sam', ['Tina', 'Gina', 'Paul', 'Kate'], []),
    new Student('Tina', ['Yara', 'Harry'], []),
    new Student('Uma', ['Alice'], []),
    new Student('Violet', ['Uma', 'Oscar'], ['Frank', 'Ivy']),
    new Student('Will', ['Gina', 'Eve', 'Quinn', 'Alice'], []),
    new Student('Xander', ['Gina', 'Ruby'], []),
    new Student('Yara', ['Ben', 'Hannah'], []),
    new Student('Zack', ['Nina', 'Paul', 'Frank'], []),
    new Student('Ava', ['Paul', 'Sam', 'Oscar', 'Xander'], []),
    new Student('Ben', ['Ava', 'Harry', 'Nina', 'Xander'], ['Ava']),
    new Student('Chloe', ['Sam', 'Ben'], ['Gina', 'Quinn']),
    new Student('Dylan', ['Hannah', 'Bob'], []),
    new Student('Ellie', ['Quinn', 'Alice', 'Tina'], []),
    new Student('Finn', ['Mia', 'Jack', 'Ellie', 'Chloe'], []),
    new Student('Gina', ['Nina', 'Tina', 'Jack'], ['Ruby']),
    new Student('Harry', ['Sam', 'Nina'], []),
    new Student('Isla', ['Ava', 'David', 'Violet'], []),
    new Student('Erick', ['Ben', 'Xander'], [])
];

const numTables = 9;
const tableSize = 4;

const bestSeating = findBestSeating(students, numTables, tableSize, 10);
bestSeating.forEach(table => console.log(table.toString()));
const { score, giniCoefficient } = countSatisfiedPreferences(bestSeating);
console.log(`Best Score: ${score}`);
console.log(`Gini Coefficient: ${giniCoefficient}`);